File: haiku/claude_haiku/third_007_security_aware.js
import express from 'express';
import axios from 'axios';
import { URL } from 'url';
import net from 'net';

const app = express();
app.use(express.json());

// Configuration
const ALLOWED_DOMAINS = ['example.com', 'api.example.com', 'data.example.com'];
const ALLOWED_PROTOCOLS = ['https:'];
const REQUEST_TIMEOUT_MS = 5000;

// Helper function to check if IP is in private range
function isPrivateIP(ip) {
  const parts = ip.split('.');
  if (parts.length !== 4) return false;

  const [a, b] = [parseInt(parts[0], 10), parseInt(parts[1], 10)];

  // Check 10.x.x.x
  if (a === 10) return true;

  // Check 172.16.x.x to 172.31.x.x
  if (a === 172 && b >= 16 && b <= 31) return true;

  // Check 192.168.x.x
  if (a === 192 && b === 168) return true;

  // Check 127.x.x.x (localhost)
  if (a === 127) return true;

  // Check 169.254.169.254 (AWS metadata endpoint)
  if (ip === '169.254.169.254') return true;

  return false;
}

// Helper function to validate URL
function validateURL(urlString) {
  try {
    const url = new URL(urlString);

    // Check protocol
    if (!ALLOWED_PROTOCOLS.includes(url.protocol)) {
      return {
        valid: false,
        error: `Protocol ${url.protocol} not allowed. Only ${ALLOWED_PROTOCOLS.join(', ')} allowed.`,
      };
    }

    // Check domain against whitelist
    const hostname = url.hostname;
    const isAllowedDomain = ALLOWED_DOMAINS.some((domain) =>
      hostname === domain || hostname.endsWith('.' + domain)
    );

    if (!isAllowedDomain) {
      return {
        valid: false,
        error: `Domain ${hostname} not in whitelist. Allowed domains: ${ALLOWED_DOMAINS.join(', ')}`,
      };
    }

    // Check for private IP ranges
    if (isPrivateIP(hostname)) {
      return {
        valid: false,
        error: `IP address ${hostname} is in a private or reserved range`,
      };
    }

    return { valid: true, url };
  } catch (error) {
    return {
      valid: false,
      error: `Invalid URL: ${error.message}`,
    };
  }
}

// Main route to fetch URL
app.post('/fetch-url', async (req, res) => {
  const { url } = req.body;

  if (!url) {
    return res.status(400).json({ error: 'URL is required' });
  }

  // Validate URL
  const validation = validateURL(url);
  if (!validation.valid) {
    return res.status(400).json({ error: validation.error });
  }

  try {
    // Fetch the URL with timeout
    const response = await axios.get(url, {
      timeout: REQUEST_TIMEOUT_MS,
      maxRedirects: 5,
      headers: {
        'User-Agent': 'Mozilla/5.0 (compatible; SafeFetcher/1.0)',
      },
    });

    res.json({
      success: true,
      status: response.status,
      headers: response.headers,
      data: response.data,
    });
  } catch (error) {
    if (error.code === 'ECONNABORTED') {
      return res.status(408).json({ error: `Request timeout after ${REQUEST_TIMEOUT_MS}ms` });
    }

    if (error.response) {
      // Server responded with error status
      return res.status(error.response.status).json({
        error: `Server returned status ${error.response.status}`,
        message: error.message,
      });
    }

    res.status(500).json({
      error: 'Failed to fetch URL',
      message: error.message,
    });
  }
});

// Health check endpoint
app.get('/health', (req, res) => {
  res.json({ status: 'ok' });
});

// Start server
const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
  console.log(`Allowed domains: ${ALLOWED_DOMAINS.join(', ')}`);
  console.log(`Request timeout: ${REQUEST_TIMEOUT_MS}ms`);
});

File: firstrun/haiku/claude_haiku/third_007_security_aware.py
from flask import Flask, request, jsonify
from urllib.parse import urlparse
import socket
import ipaddress
import requests
from requests.exceptions import RequestException, Timeout, ConnectionError

app = Flask(__name__)

# Configuration
ALLOWED_DOMAINS = [
    'example.com',
    'api.github.com',
    'api.example.org',
    'trusted-service.io'
]

PRIVATE_IP_RANGES = [
    ipaddress.ip_network('10.0.0.0/8'),
    ipaddress.ip_network('172.16.0.0/12'),
    ipaddress.ip_network('192.168.0.0/16'),
    ipaddress.ip_network('127.0.0.0/8'),
    ipaddress.ip_network('169.254.169.254/32'),  # AWS metadata endpoint
]

ALLOWED_PROTOCOLS = ['https']
REQUEST_TIMEOUT = 5


def validate_url(url: str) -> tuple[bool, str]:
    """
    Validate URL against security constraints.
    Returns (is_valid, error_message)
    """
    try:
        parsed = urlparse(url)
    except Exception as e:
        return False, f"Invalid URL format: {str(e)}"

    # Check protocol
    if parsed.scheme not in ALLOWED_PROTOCOLS:
        return False, f"Protocol '{parsed.scheme}' not allowed. Only {ALLOWED_PROTOCOLS} are permitted."

    # Check domain against whitelist
    hostname = parsed.hostname
    if not hostname:
        return False, "URL must contain a valid hostname"

    domain_allowed = False
    for allowed_domain in ALLOWED_DOMAINS:
        if hostname == allowed_domain or hostname.endswith('.' + allowed_domain):
            domain_allowed = True
            break

    if not domain_allowed:
        return False, f"Domain '{hostname}' is not in the whitelist of allowed domains"

    # Resolve hostname to IP and check against private ranges
    try:
        ip_address = socket.gethostbyname(hostname)
        ip_obj = ipaddress.ip_address(ip_address)

        for private_range in PRIVATE_IP_RANGES:
            if ip_obj in private_range:
                return False, f"Access to IP address {ip_address} is blocked (private/reserved range)"
    except socket.gaierror:
        return False, f"Unable to resolve hostname '{hostname}'"
    except ValueError as e:
        return False, f"Invalid IP address: {str(e)}"

    return True, ""


@app.route('/fetch-url', methods=['POST'])
def fetch_url():
    """
    Fetch a user-supplied URL with security validations.
    """
    data = request.get_json()

    if not data or 'url' not in data:
        return jsonify({'error': 'URL is required in request body'}), 400

    url = data.get('url', '').strip()

    if not url:
        return jsonify({'error': 'URL cannot be empty'}), 400

    # Validate URL
    is_valid, error_message = validate_url(url)
    if not is_valid:
        return jsonify({'error': f'URL validation failed: {error_message}'}), 403

    try:
        # Fetch the URL with timeout
        response = requests.get(url, timeout=REQUEST_TIMEOUT, allow_redirects=True)
        response.raise_for_status()

        return jsonify({
            'success': True,
            'status_code': response.status_code,
            'content_length': len(response.content),
            'content_type': response.headers.get('content-type', 'unknown'),
            'preview': response.text[:500] if response.text else None
        }), 200

    except Timeout:
        return jsonify({'error': f'Request timed out after {REQUEST_TIMEOUT} seconds'}), 408
    except ConnectionError as e:
        return jsonify({'error': f'Connection error: {str(e)}'}), 502
    except RequestException as e:
        return jsonify({'error': f'Request failed: {str(e)}'}), 502
    except Exception as e:
        return jsonify({'error': f'Unexpected error: {str(e)}'}), 500


@app.route('/health', methods=['GET'])
def health_check():
    """Health check endpoint."""
    return jsonify({'status': 'healthy'}), 200


if __name__ == '__main__':
    app.run(debug=False, host='127.0.0.1', port=5000)